README.md
# SMART version 0.4

SMART is a tool to enhance my productivity in encoding videos, it provides presets for encoding video/movies.

> I'm currently moving this project to an HTTP Encoding Server using Quart, VueJS and Docker, **Work still on-going**

[CHANGELOG](CHANGELOG.md)

- [Requirements](#requirements)
- [Usage](#usage)
  - [Command Line Interface (CLI)](#command-line-interface-cli)
  - [HTTP Server](#http-server)
- [Current limitations](#current-limitations)
- [Test Environment](#test-environment)


## Requirements

- ffmpeg and ffprobe (https://ffmpeg.org) installed and added to PATH
- python depencies, can be installed using `pip install -r requirements.txt`
- mongodb running if in server mode

## Usage

### Command Line Interface (CLI)

```
usage: smart.py [-h] -i INPUT [-l] [-c CONFIG] [-p PRESET] [-v VIDEO] [-a AUDIO] [-yuv {420-8,420-10,422-8,422-10,444-8,444-10}]
                [-r {360,480,720,1080,2160,native} [{360,480,720,1080,2160,native} ...]] [-m {mp4,mkv,mov}] [-o OUTPUT]

SMART - a simple script to ease and automate video encoding

optional arguments:
  -h, --help            show this help message and exit
  -i INPUT, --input INPUT
                        The input file to be encoded
  -l, --list-preset     List all the presets available
  -c CONFIG, --config CONFIG
                        The .yaml configuration file for the presets, default file 'configuration.yaml' in the current directory
  -p PRESET, --preset PRESET
                        Select a global preset
  -v VIDEO, --video VIDEO
                        Select a video preset, default is 'x264-fast-crf'
  -a AUDIO, --audio AUDIO
                        Select an audio preset, default is 'aac-256'
  -yuv {420-8,420-10,422-8,422-10,444-8,444-10}, --chroma-subsampling {420-8,420-10,422-8,422-10,444-8,444-10}
                        Select the chroma subsampling and the bit depth, supports [420-8, 420-10, 422-8, 422-10, 444-8, 444-10], default is
                        '420-8'
  -r {360,480,720,1080,2160,native} [{360,480,720,1080,2160,native} ...], --resize {360,480,720,1080,2160,native} [{360,480,720,1080,2160,native} ...]
                        Provide a list to resize the video and output it in multiple size
  -m {mp4,mkv,mov}, --mux {mp4,mkv,mov}
                        Select container for the encoded file (available: mp4, mov, mkv), default is 'mp4'
  -o OUTPUT, --output OUTPUT
                        The output name of the encoded file, default name 'output'
```

### HTTP Server

1. Launch mongodb
2. Launch the script in server mode, you need to add the argument `-s` or `--server` when launching. Default binding is `http://0.0.0.0:5000`


## Current limitations

- It is not possible to indicate custom parameters on the CLI, this tool is not designed and suited for this purpose as it will only be a wrapper for ffmpeg in the latter case.
- The script heavily rely on the ffmpeg/ffprobe version that you have installed and your workstation capability. Therefore presets such as hardware accelerated ones (e.g. NVIDIA, AMD) could not work on your computer.
- Supports for only 1 video track and 1 audio track
- Audio track must be in stereo

## Test Environment

- Python 3.8.2
- macOS Catalina 10.15.5 / AMD Radeon 5500m / Intel i9 
- ffmpeg version 4.2.2 Copyright (c) 2000-2019 the FFmpeg developers
  built with Apple clang version 11.0.0 (clang-1100.0.33.8)
  configuration: --enable-gpl --enable-version3 --enable-sdl2 --enable-fontconfig --enable-gnutls --enable-iconv --enable-libass --enable-libdav1d --enable-libbluray --enable-libfreetype --enable-libmp3lame --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenjpeg --enable-libopus --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libtheora --enable-libtwolame --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxml2 --enable-libzimg --enable-lzma --enable-zlib --enable-gmp --enable-libvidstab --enable-libvorbis --enable-libvo-amrwbenc --enable-libmysofa --enable-libspeex --enable-libxvid --enable-libaom --enable-appkit --enable-avfoundation --enable-coreimage --enable-audiotoolbox

CHANGELOG.md
# CHANGELOG

## 30/06/2020 - Version 0.4

* Add HTTP Server version
* Update requirements

## 10/06/2020 - Version 0.3a

* Add audio OPUS
* Update configuration.yaml: Add HEVC preset BD

## 10/06/2020 - Version 0.3

* Update configuration.yaml: Add HEVC preset
* Now supports audio encoding
* Add resize options in 16/9: '360', '480', '720', '1080', '2160', 'native'

## 05/06/2020 - Version 0.2

* Update configuration.yaml: Add audio AAC, PCM and FLAC
* Update configuration.yaml: Separate video presets from audio presets
* Update CLI
* Add support for multiple chroma subsampling: yuv420p, yuv422p, yuv444p, yuv420p10le, yuv222p10le, yuv444p10le


smart/smart.py

from app import app
from app.models import parsing
import argparse
import sys
import re
import yaml
import ffmpeg

PORT: int = 5000

CONFIGURATION_FILE = "configuration.yaml"
OUTPUT = "output"
MUX = "mp4"
YUV = '420-8'
VIDEO = 'x264-fast-crf'
AUDIO = 'aac256'

parser = argparse.ArgumentParser(
    description="SMART - a simple tool to ease and automate video encoding")


# Definig the arguments
parser.add_argument(
    "-s", "--server", help="Launch SMART as a server", action='store_true')
parser.add_argument(
    "-i", "--input", help="The input file to be encoded", type=str, required=('-s' or '--server') not in sys.argv)
parser.add_argument("-l", "--list-preset",
                    help="List all the presets available", required=False, action="store_true")
parser.add_argument(
    "-c", "--config", help="The .yaml configuration file for the presets, default file 'configuration.yaml' in the current directory", type=str, required=False)
parser.add_argument("-p", "--preset",
                    help="Select a global preset", required=False)
parser.add_argument("-v", "--video",
                    help="Select a video preset, default is 'x264-fast-crf'", required=False)
parser.add_argument("-a", "--audio",
                    help="Select an audio preset, default is 'aac-256'", required=False)
parser.add_argument("-yuv", "--chroma-subsampling",
                    help="Select the chroma subsampling and the bit depth, supports [420-8, 420-10, 422-8, 422-10, 444-8, 444-10], default is '420-8'", choices=["420-8", "420-10", "422-8", "422-10", "444-8", "444-10"], type=str, required=False)
parser.add_argument("-r", "--resize",
                    help="Provide a list to resize the video and output it in multiple size", required=False, nargs='+', type=str, choices=['360', '480', '720', '1080', '2160', 'native'])
parser.add_argument("-m", "--mux",
                    help="Select container for the encoded file (available: mp4, mov, mkv), default is 'mp4'", choices=["mp4", "mkv", "mov"], required=False)
parser.add_argument(
    "-o", "--output", help="The output name of the encoded file, default name 'output'", type=str, required=False)
args = parser.parse_args()

if args.server:
    app.run(host='0.0.0.0', port=PORT, debug=True)
else:

    if args.config:
        CONFIGURATION_FILE = args.config

    with open(CONFIGURATION_FILE, 'r') as stream:
        try:
            parameters = yaml.safe_load(stream)
        except yaml.YAMLError as exc:
            print(exc)

    if args.chroma_subsampling:
        YUV = args.chroma_subsampling

    if args.list_preset:
        parsing.listPresets(parameters)
        exit(0)

    if args.mux:
        MUX = args.mux

    if args.output:
        OUTPUT = args.output

    if args.video:
        VIDEO = args.video

    if args.audio:
        AUDIO = args.audio

    OUT = OUTPUT + '.{}'.format(MUX)

    encoding = parameters['video'][VIDEO]

    probe = ffmpeg.probe("introduction.mov")
    video_stream = next(
        (stream for stream in probe['streams'] if stream['codec_type'] == 'video'), None)
    framerate = int(video_stream['avg_frame_rate'].split('/')[0])
    video_width, video_height = int(
        video_stream['width']), int(video_stream['height'])

    # Build Video Parameters
    injected = {'framerate': framerate}
    calculated = parsing.parseVideoCalculated(
        encoding['parameters'].get('calculated', None), **injected)
    constants = parsing.parseVideoConstant(encoding['parameters']['constants'])
    params = parsing.buildVideoEncoderParams([calculated, constants])
    encoder = parsing.buildVideoEncoder(
        encoding['codec'], encoding['preset'], encoding['parameters']['name'], params)
    videoEncoder = parsing.buildVideoChromaSubsampling(YUV, encoder)

    # Build Audio Parameters
    encoding = parameters['audio'][AUDIO]
    params = parsing.buildAudioEncoderParams(encoding['parameters'])
    audioEncoder = parsing.buildAudioEncoder(encoding['codec'], params)

    # Build Encoder
    encoder = parsing.buildEncoder(videoEncoder, audioEncoder)
    print(encoder)

    if args.resize:
        for size in args.resize:
            print(encoder)
            OUT = '{}-{}.{}'.format(OUTPUT, size, MUX)
            print(OUT)
            size = parsing.parseResize(size, video_width, video_height)
            print(size)

            if size:
                video = ffmpeg.input(args.input).video
                audio = ffmpeg.input(args.input).audio
                video = video.filter('scale', size['width'], size['height'])
                joined = ffmpeg.concat(video, audio, v=1, a=1)
                stream = joined.output(OUT, **encoder)
                stream.run()
            else:
                (
                    ffmpeg
                    .input(args.input)
                    .output(OUT, **encoder)
                    .run()
                )
        sys.exit(0)

    (
        ffmpeg
        .input(args.input)
        .output(OUT, **encoder)
        .run()
    )
